### ISSUE TITLE
Weird problem with username validation "The user name you entered is reserved. It cannot be used."

### ISSUE THREAD

**Issue by Kaik** [zikula/core#2942]
| Q                 | A
| ----------------- | ---
| Zikula Version    | 1.4.3
| PHP Version       | 5.6

In ZikulaIntercomModule which I'm refactoring, I have a new message form and there is a recipient field.
https://github.com/Kaik/InterCom/blob/master/Form/Type/MessageType.php#L30
When I type admin as recipient and press send it shows me this error above that field ``` The user name you entered is reserved. It cannot be used.``` Since I did not implement this kind of validation I have looked in core and it is present here https://github.com/zikula/core/blob/02ec0226f7d18196d2cab099f2ab2b26327b1fdd/src/system/ZAuthModule/Validator/Constraints/ValidUnameValidator.php#L84
and here
https://github.com/zikula/core/blob/02ec0226f7d18196d2cab099f2ab2b26327b1fdd/src/system/UsersModule/Validator/Constraints/ValidUnameValidator.php#L84

### Expected behavior
UsersModule and ZAuth module username validation to work only in those modules.

### Actual behavior
UsersModule and ZAuth module username validation works globally?

### Steps to reproduce
Install master from here https://github.com/Kaik/InterCom and send a message to admin. Of course, admin username needs to be restricted in UsersModule settings

This is very wired... @craigh

**Comment by craigh** [comment]
> UsersModule and ZAuth module username validation works globally?

as I understand it, this is impossible.

**Comment by craigh** [comment]
I'm guessing here, but here's what I will guess is happening:

The validation is built into the UserEntity (see [this](https://github.com/zikula/core/blob/02ec0226f7d18196d2cab099f2ab2b26327b1fdd/src/system/UsersModule/Entity/UserEntity.php#L28)). So in order to get the error you must be (1) validating and entity and (2) attempting to persist it (I think). So, you might check some cascade behavior in your other entities. Or take a look at your logic (which I have not done). Obviously it *should not* be getting validated or persisted 😄 so the question is why is it doing that? You can also look at the search methods ([here](https://github.com/zikula/core/blob/ce9260a54d368a13351dc9a915f36313ef2df7c7/src/system/UsersModule/Controller/UserAdministrationController.php#L100-L119) and [here](https://github.com/zikula/core/blob/ce9260a54d368a13351dc9a915f36313ef2df7c7/src/system/UsersModule/Controller/UserAdministrationController.php#L345-L371) in the users module to see how it is different than yours.

IMO this is not a core problem.

**Comment by Kaik** [comment]
Hmm I will test it, but I would not be so sure that it is not core. Because I remember testing it before 1.4.1 or even 1.4.2 well about 6-9 months ago and it was working. There is only one entity messages with
```
    /**
     * The recipient uid
     *     
     * @ORM\ManyToOne(targetEntity="Zikula\Module\UsersModule\Entity\UserEntity")
     * @ORM\JoinColumn(name="recipient", referencedColumnName="uid")
     */
    private $recipient;
```
There are things I suspect could be a reason as well like ModelTransformer name that I use gets overridden somewhere and interfere with this validation, but for me most probable is that this validation somehow works globally... anyway I will trace it

**Comment by craigh** [comment]
despite the fact that I did so in Dizkus, I can no longer recommend non-core modules using a `UserEntity` as a relation as you doing here. I would recommend storing only the UID.

**Comment by Kaik** [comment]
What wonders me more is that MessageEntity has two user fields sender and recipient and only recipient gives this error, sending message from admin account to other account works only when I try to send it to admin account triggers this error.

**Comment by Kaik** [comment]
What you mean we should store it as integer not as relation?

**Comment by craigh** [comment]
> sending message from admin account to other account works

then there is a logic error somewhere.

> What you mean should I store it as integer not as relation?

/**
     * The recipient uid
     *     
     * @ORM\Column(type="integer")
     */
    private $recipient;

**Comment by Kaik** [comment]
So what you are saying is to leave the idea of related objects on which Doctrine base on and use integer instead. So module developers to populate user object will have to either use join in the repository or I do not know iterate over objects array to get user object by id if they need it? This is so bad... I do not understand why it is wrong to use a relation to ``` UserEntity``` this is like back to the stone age. User object for any entity is the most wanted one please check any module table and there is at least one user field. I really would like to know what is the reason behind this. It is like you would say let's not use relations at all and do everything the old way using id's and joins...

**Comment by craigh** [comment]
holy smokes. overreact much? take a breath.

what I'm saying is that if your problem works sometimes and not others, then it is a logic problem.

**Comment by cmfcmf** [comment]
> holy smokes. overreact much? take a breath.

I am tempted to agree with @Kaik though.

**Comment by Kaik** [comment]
Lol I dissagre that it is logic problem well it might be but not in IntercomModule. What I pointed out is that it lays somewhere in form layer it does not work only with model transformer where I type user name and then convert it to id recipient field is done this way while sender is internal by uid and it works. I will be back soon I'm writing from phone

**Comment by craigh** [comment]
Please test your issue against latest core changes.

**Comment by Kaik** [comment]
I will tomorrow. I'm making changes in my office (new bench) and computer on which I'm working is... well it is on the floor right now in pieces.
I have seen changes you have made and I do not think it will help as it works only for admins. 
What would help is to restrict validation only to entity creation, but it would be kind of wrong as well.

I'm very surprised that this error is triggered anyway - as I'm not changing user entity only "selecting it" in many to one relation. I was looking for any similar usage (where there is DataTransformer and Validation Constraint) but I could not find any.

I had no time to investigate it further I was really busy last month. What I have found some time ago is that there are two kinds of transformers Model and View I'm using model transformer now. It used to work in core 1.4.2, but it might be that I need to use View transformer instead of Model both are very similar but are woking on different levels and it might work I need to test it. http://symfony.com/doc/current/form/data_transformers.html#about-model-and-view-transformers

Btw my new bench 
![imag0285](https://cloud.githubusercontent.com/assets/966910/17603849/472d125a-6013-11e6-8915-ba03add98ba8.jpg)

**Comment by craigh** [comment]
> I do not think it will help as it works only for admins. 

there are other changes and therefore my request.

**Comment by Kaik** [comment]
Ok I will check it tomorrow.

**Comment by Kaik** [comment]
This issue is still present I will now check maybe I can get view transformer to work.
![screenshot from 2016-08-14 11-02-47](https://cloud.githubusercontent.com/assets/966910/17648283/2f3940b2-620f-11e6-9f4c-652d71c49dbb.png)

**Comment by Kaik** [comment]
This is still the same. Core at  4961ffa...
I have tested both ``` addModelTransformer ``` and ``` addViewTransformer ```
![screenshot from 2016-08-14 11-02-47](https://cloud.githubusercontent.com/assets/966910/17648345/09a6791c-6211-11e6-9bc7-23c0c8c26e8c.png)

I have even disabled "saving" so I can see where error occurs and it occurs during validation.

```
   public function newMessageAction(Request $request) {
        // Permission check
        if (!$this->get('zikula_intercom_module.access_manager')->hasPermission()) {
            throw new AccessDeniedException();
        }
        $options = ['isXmlHttpRequest' => $request->isXmlHttpRequest()];
        $message = $this->get('zikula_intercom_module.manager.message')->create();
        $form = $this->createForm('messageform', $message, $options);
        $form->handleRequest($request);

        if ($form->isValid()) {
//            $this->get('zikula_intercom_module.manager.message')->setNewData($form->getData());
//            $this->get('zikula_intercom_module.manager.message')->send();
//            if ($request->isXmlHttpRequest()) {
//                return new JsonResponse(array('status' => true));
//            } else {
//                $request->getSession()
//                        ->getFlashBag()
//                        ->add('status', "Message send!");
//                return $this->redirect($this->generateUrl('zikulaintercommodule_inbox_view'));
//            }
        }

        if ($request->isXmlHttpRequest()) {
           return new JsonResponse(array('status' => true, 'html' => $this->renderView('ZikulaIntercomModule:Message:form.html.twig', array(
                    'form' => $form->createView(),
                    'message' => $message,
                    'settings' => $this->getVars()
        )))); 
        }
        $request->attributes->set('_legacy', true); // forces template to render inside old theme
        return $this->render('ZikulaIntercomModule:Message:new.html.twig', array(
                    'form' => $form->createView(),
                    'message' => $message,
                    'settings' => $this->getVars()
        ));
    } 
```

```
    public function buildForm(FormBuilderInterface $builder, array $options) {

        $builder->setMethod('POST')
                ->add($builder->create('recipient', 'text', ['invalid_message' => $this->__('User not found.'),'attr' => ['class' => 'author_search']])
                        ->addModelTransformer(new UserToIdTransformer($this->entityManager)))
                ->add($builder->create('groups', 'text', ['mapped' => false,'required' => false,'invalid_message' => $this->__('Group not found.'), 'attr' => ['class' => 'author_search']])
                        ->addModelTransformer(new GroupToIdTransformer($this->entityManager)))
                ->add('subject', 'text', [
                    'required' => false
                ])
                ->add('text', 'textarea', [
                    'required' => true, 'attr' => [
                        'class' => 'tinymce'
                    ]
                ]);
   
        if ($options['isXmlHttpRequest'] == false) {
            $builder->add('save', 'submit', [
                'label' => $this->__('Send'),
                'attr' => [
                    'class' => 'btn-success'
                ]
            ]);
        }       

    }
```

**Comment by Kaik** [comment]
Here I have found one information regarding this behaviour, well it is oposite situation but looks like there is answer what is causing validation to occur before transformation. http://www.scriptscoop.me/t/fe4a2b8398a6/forms-symfony2-form-field-constraints-validation-before-data-transformer.html

**Comment by craigh** [comment]
can you mail me the module so I can see it for myself?

**Comment by Kaik** [comment]
Yes, I will do it today later one I'm in the middle of working on it and some parts are disabled etc so I need to test if it installs properly.

Btw, I just wanted to check if adding constraint validation to form would work I have checked forms and I was surprised to find out that it is already implemented there, so I went to UserEntity and removed this line https://github.com/zikula/core/blob/1.4/src/system/UsersModule/Entity/UserEntity.php#L47 and everything works both validation in registration form and sending message to admin.

I will check if module installs properly and will email it to you today.
